lib/utils/mailup.ts

```diff
@@ -0,0 +1,287 @@
+const TRANSPORT_VERSION = "0.1";
+
+/**
+ * You need to create a SMTP+ user in MailUp administration panel
+ */
+export const SmtpAuthInfo = t.interface({
+  Secret: NonEmptyString,
+  Username: NonEmptyString
+});
+
+export type SmtpAuthInfo = t.TypeOf<typeof SmtpAuthInfo>;
+
+/**
+ * MailUp API calls common response fields
+ */
+const ApiResponse = t.interface({
+  Code: t.string,
+  Message: t.string,
+  Status: t.string
+});
+
+type ApiResponse = t.TypeOf<typeof ApiResponse>;
+
+const Address = t.interface({
+  Email: EmailString,
+  Name: t.string
+});
+
+type Address = t.TypeOf<typeof Address>;
+
+const NameValue = t.interface({
+  N: NonEmptyString,
+  V: t.string
+});
+
+type NameValue = t.TypeOf<typeof NameValue>;
+
+const Html = t.interface({
+  Body: NonEmptyString
+});
+
+type Html = t.TypeOf<typeof NameValue>;
+
+const EmailPayload = t.intersection([
+  t.interface({
+    ExtendedHeaders: t.array(NameValue),
+    From: Address,
+    Html,
+    Subject: NonEmptyString,
+    Text: NonEmptyString,
+    To: t.array(Address)
+  }),
+  t.partial({
+    Bcc: t.array(Address),
+    Cc: t.array(Address),
+    ReplyTo: t.string
+  })
+]);
+
+type EmailPayload = t.TypeOf<typeof EmailPayload>;
+
+export interface IMailUpTransportOptions {
+  readonly creds: SmtpAuthInfo;
+}
+
+interface IAddresses {
+  readonly bcc?: ReadonlyArray<NodemailerAddress>;
+  readonly cc?: ReadonlyArray<NodemailerAddress>;
+  readonly from?: ReadonlyArray<NodemailerAddress>;
+  readonly sender?: ReadonlyArray<NodemailerAddress>;
+  readonly "reply-to"?: ReadonlyArray<NodemailerAddress>;
+  readonly to?: ReadonlyArray<NodemailerAddress>;
+}
+
+function callMailUpApi(
+  method: HttpMethod,
+  url: string,
+  creds: SmtpAuthInfo,
+  payload: {}
+): Promise<Either<Error, ApiResponse>> {
+  return request(method, url)
+    .timeout(DEFAULT_REQUEST_TIMEOUT_MS)
+    .send({ ...payload, User: creds })
+    .retry(DEFAULT_REQUEST_RETRIES)
+    .then(response => {
+      if (response.error) {
+        return left<Error, ApiResponse>(
+          new Error(
+            `Error calling MailUp API: ${response.status} - ${response.text}`
+          )
+        );
+      }
+      return right<Error, ApiResponse>(response.body);
+    })
+    .catch(err => left<Error, ApiResponse>(new Error(err.response)));
+}
+
+async function sendTransactionalMail(
+  creds: SmtpAuthInfo,
+  payload: EmailPayload
+): Promise<Either<Error, ApiResponse>> {
+  const errorOrResponse = await callMailUpApi(
+    "POST",
+    ENDPOINTS.sendTransactionalMail,
+    creds,
+    payload
+  );
+  if (isLeft(errorOrResponse)) {
+    return errorOrResponse;
+  } else {
+    const response = errorOrResponse.value;
+    if (response && response.Code && response.Code === "0") {
+      return right(response);
+    } else {
+      return left(
+        new Error(
+          `Error while sending email: ${response.Code} - ${response.Message}`
+        )
+      );
+    }
+  }
+}
+
+function toMailupAddresses(
```

Comment:
not super clear what this function and the next do, I would refactor them a little bit to make it easier to understand the validation logic - also functions that accept `undefined` values smell of problems (it's usually the caller that has to check the value and decide whether to call this function or not)